internal/command/command_service.go

```diff
@@ -0,0 +1,286 @@
+		}
+
+		response, updateError := cs.commandServiceClient.UpdateDataPlaneHealth(ctx, request)
+
+		validatedError := grpc.ValidateGrpcError(updateError)
+
+		if validatedError != nil {
+			slog.ErrorContext(ctx, "Failed to send update data plane health", "error", validatedError)
+
+			return nil, validatedError
+		}
+
+		return response, nil
+	}
+
+	response, err := backoff.RetryWithData(
+		sendDataPlaneHealth,
+		backoffHelpers.Context(backOffCtx, cs.agentConfig.Common),
+	)
+	if err != nil {
+		return err
+	}
+
+	slog.DebugContext(ctx, "UpdateDataPlaneHealth response", "response", response)
+
+	return err
+}
+
+func (cs *CommandService) CancelSubscription(ctx context.Context) {
+	slog.InfoContext(ctx, "Canceling subscribe context")
+
+	cs.subscribeMutex.Lock()
+	if cs.subscribeCancel != nil {
+		cs.subscribeCancel()
+	}
+	cs.subscribeMutex.Unlock()
+}
+
+// nolint: revive,gocognit
+func (cs *CommandService) subscribe(ctx context.Context) {
+	var subscribeClient mpi.CommandService_SubscribeClient
+	var err error
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			subscribeFn := func() error {
+				if subscribeClient == nil {
+					if cs.commandServiceClient == nil {
+						return errors.New("command service client is not initialized")
+					}
+
+					subscribeClient, err = cs.commandServiceClient.Subscribe(ctx)
+					if err != nil {
+						slog.ErrorContext(ctx, "Failed to create subscribe client", "error", err)
+
+						return err
+					}
+
+					if subscribeClient == nil {
+						return errors.New("subscribe client is not initialized")
+					}
+				}
+
+				request, recvError := subscribeClient.Recv()
+				if recvError != nil {
+					slog.ErrorContext(ctx, "Failed to receive message from subscribe stream", "error", recvError)
+					subscribeClient = nil
+
+					return recvError
+				}
+
+				cs.subscribeChannel <- request
+
+				return nil
+			}
+
+			commonSettings := &config.CommonSettings{
+				InitialInterval:     cs.agentConfig.Common.InitialInterval,
+				MaxInterval:         cs.agentConfig.Common.MaxInterval,
+				MaxElapsedTime:      createConnectionMaxElapsedTime,
+				RandomizationFactor: cs.agentConfig.Common.RandomizationFactor,
+				Multiplier:          cs.agentConfig.Common.Multiplier,
+			}
+
+			retryError := backoff.Retry(subscribeFn, backoffHelpers.Context(ctx, commonSettings))
+			if retryError != nil {
+				slog.ErrorContext(ctx, "Failed to receive messages from subscribe stream", "error", retryError)
+			}
+		}
+	}
+}
+
+func (cs *CommandService) createConnection(ctx context.Context, resource *mpi.Resource) error {
+	correlationID := logger.GetCorrelationID(ctx)
+
+	request := &mpi.CreateConnectionRequest{
+		MessageMeta: &mpi.MessageMeta{
+			MessageId:     uuid.NewString(),
+			CorrelationId: correlationID,
+			Timestamp:     timestamppb.Now(),
+		},
+		Resource: resource,
+	}
+
+	connectFn := func() (*mpi.CreateConnectionResponse, error) {
```

Comment:
I'd say for readability, move this to it's own function and not inline. Can potentially have a unit test then

Reply:
cant since it needs variables that are outside of the function like `ctx` & `request`

Reply:
You can have a function that returns a function:

```go
func (cs *CommandService) createConnectionFunc(ctx context.Context, request *mpi.CreateConnectionRequest) func() (*mpi.CreateConnectionResponse, error) {
    return func() (*mpi.CreateConnectionResponse, error) {
        response, connectErr := cs.commandServiceClient.CreateConnection(ctx, request)

        validatedError := grpc.ValidateGrpcError(connectErr)
        if validatedError != nil {
             slog.ErrorContext(ctx, "Failed to create connection", "error", validatedError)

            return nil, validatedError
        }

	return response, nil
    }
}
```

Dunno if it'll get messy though